### PR TITLE
GC: two runnable steps proposal

### LINKED ISSUE #4255: GC - intermediate results
Allow examination of intermediate results before deleting the objects, and a way to proceed with the previously generated results.

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
This proposal includes breaking a GC run into two runnable steps:
1. **Marking run**- collecting the addresses to be removed.
2. **Sweeping run**- removing the objects that were mapped by the collected addresses.

Closes #4255

### CODE REVIEW

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [inline design/open/gc-two-steps-separation.md:16]: @ozkatz had a great idea to use Terraform terminology here: 
only mark -> plan 
only sweep -> apply with a plan id 
mark id -> plan id

@Jonathan-Rosenberg WDYT?

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [inline design/open/gc-two-steps-separation.md:25]: I think that we should fail GC in case this permutation is used, why would we want to override a previous GC plan run (I'm using the new terminology)? 
thoughts?

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [inline design/open/gc-two-steps-separation.md:58]: Thanks for including this, it was my next question

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [inline design/open/gc-two-steps-separation.md:70]: I'm not sure that I understand this section, can you please clarify?

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [inline design/open/gc-two-steps-separation.md:71]: We need to think about how and when to make this feature available in lakeFS cloud, and about its user experience. Since cloud users can't control the GC run command, this will require additional effort. I suggest discussing this separately. Can you please open an issue for it? 
cc @ozkatz

**Review by talSofer — APPROVED** [review]
Thanks @Jonathan-Rosenberg, this looks great! 

I'm approving but will appreciate it if you consider the comments before merging.

File: design/open/gc-two-steps-separation.md
Comment (Jonathan-Rosenberg) [reply design/open/gc-two-steps-separation.md:16]: I actually think that we should GC terminology with the GC...

File: design/open/gc-two-steps-separation.md
Comment (Jonathan-Rosenberg) [reply design/open/gc-two-steps-separation.md:25]: This is the case where we only want to mark the addresses to be deleted with a specific mark ID so we can examine the results according to it.
Why should it fail here?
As explained in the [Note](https://github.com/treeverse/lakeFS/pull/4254/files#diff-5e7e580bed0e5e9fdb5a1a66fab9a832d7229cbf7ed34596b0a5f22c86ff6eb4R53) below, the users who orchestrate the GC run will need to ensure that they don't override the previous mark/plan or do override it if they wish...

Comment (arielshaqed) [reply design/open/gc-two-steps-separation.md:25]: It is a very useful flag combination.  I can use it for automation or testing or whatever.

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [reply design/open/gc-two-steps-separation.md:16]: Ok, I get it. makes sense

Comment (arielshaqed) [reply design/open/gc-two-steps-separation.md:16]: I prefer mark/sweep which is established terminology for this.

Neither mark/sweep nor plan/apply make it entirely clear that the second stage will delete stuff.  But sweep is clearer, that's another point in its favour.

File: design/open/gc-two-steps-separation.md
Comment (talSofer) [reply design/open/gc-two-steps-separation.md:25]: I'm ok with it, thanks

File: design/open/gc-two-steps-separation.md
Comment (Jonathan-Rosenberg) [reply design/open/gc-two-steps-separation.md:16]: cool
I'll change it

**Pull request merged**
